### PR TITLE
CUDA driver fixes

### PR DESCRIPTION
**Pull request by franz**
Various fixes to make chipStar work with the PoCL-CUDA driver
- change compilation workflow to per-program CUDA module instead of per-kernel CUDA module, this allows program scope variables
- OpenCL 2.x atomics
- Subgroup support (partial)
- enable FP16 & generic AS support (with SPIR-V input)

### CODE REVIEW

File: lib/CL/devices/cuda/pocl-cuda.c [hunk 1]
```diff
@@ -124,6 +147,27 @@ typedef struct pocl_cuda_event_data_s
...
+  int supports_cu_mem_host_register;
+  int supports_managed_memory;
+  int sm_maj, sm_min, warp_size;
+  cl_bool available;
+
+  pocl_cuda_kernel_data_t CudaBuiltinKernelsData[CUDA_BUILTIN_KERNELS];
+  pocl_cuda_kernel_data_t CudnnBuiltinKernelsData[CUDNN_BUILTIN_KERNELS];
+  pocl_cuda_program_data_t CudaBuiltinKernelsProgram;
+  pocl_cuda_program_data_t CudnnBuiltinKernelsProgram;
+  int CudaBuiltinKernelsBuilt;
```
Comment (pjaaskel) [inline lib/CL/devices/cuda/pocl-cuda.c:167]: C style

File: lib/CL/devices/cuda/pocl-cuda.c [hunk 2]
```diff
@@ -581,13 +542,53 @@ pocl_cuda_init (unsigned j, cl_device_id dev, const char *parameters)
...
+
+  /* OpenCL 3.0 properties */
+  /* Minimum mandated capability */
+  dev->atomic_memory_capabilities
+      = CL_DEVICE_ATOMIC_ORDER_RELAXED | CL_DEVICE_ATOMIC_SCOPE_WORK_GROUP;
+  dev->atomic_fence_capabilities = CL_DEVICE_ATOMIC_ORDER_RELAXED
+                                   | CL_DEVICE_ATOMIC_ORDER_ACQ_REL
+                                   | CL_DEVICE_ATOMIC_SCOPE_WORK_GROUP;
+
+  // TODO should check from Compute Capability ??
```
Comment (pjaaskel) [inline lib/CL/devices/cuda/pocl-cuda.c:562]: Yep. I think safest to not assume this by default.

File: lib/CL/devices/cuda/pocl-cuda.c [hunk 3]
```diff
@@ -581,13 +542,53 @@ pocl_cuda_init (unsigned j, cl_device_id dev, const char *parameters)
...
   // https://forums.developer.nvidia.com/t/max-size-of-cuda-arguments/50218
   dev->max_parameter_size = 4352;
 
-  dev->data = data;
+#if (CUDA_DEVICE_CL_VERSION_MAJOR >= 3)
+  dev->features = CUDA_DEVICE_FEATURES_30;
+  /* this is not enabled because we're compiling program.bc and
+   * handing it over to CUDA; we're not running workgroup function
+   * generation. Therefore CUDA takes care of program-scope variables. */
+  dev->program_scope_variables_pass = CL_FALSE;
```
Comment (pjaaskel) [inline lib/CL/devices/cuda/pocl-cuda.c:578]:
```suggestion
  dev->run_program_scope_variables_pass = CL_FALSE;
```

File: lib/CL/devices/cuda/pocl-cuda.c [hunk 4]
```diff
@@ -776,20 +783,41 @@ pocl_cuda_alloc_mem_obj (cl_device_id device, cl_mem mem, void *host_ptr)
     {
       result = cuMemHostAlloc (&p->extra_ptr, mem->size,
                                CU_MEMHOSTREGISTER_DEVICEMAP);
-      CUDA_CHECK (result, "cuMemHostAlloc");
+      if (CUDA_CHECK_ERROR (result, "cuMemHostAlloc"))
+        {
+          p->extra_ptr = NULL;
```
Comment (pjaaskel) [inline lib/CL/devices/cuda/pocl-cuda.c:788]: Is this after clang-format? The indentation seems to small.

File: lib/CL/devices/cuda/pocl-cuda.c [hunk 5]
```diff
@@ -885,7 +908,7 @@ pocl_cuda_submit_read (CUstream stream, void *host_ptr, const void *device_ptr,
   POCL_MSG_PRINT_CUDA ("cuMemcpyDtoHAsync %p -> %p / %zu B \n", device_ptr, host_ptr, cb);
   CUresult result = cuMemcpyDtoHAsync (
       host_ptr, (CUdeviceptr) (device_ptr + offset), cb, stream);
-  CUDA_CHECK (result, "cuMemcpyDtoHAsync");
+  CUDA_CHECK_ABORT (result, "cuMemcpyDtoHAsync");
```
Comment (pjaaskel) [inline lib/CL/devices/cuda/pocl-cuda.c:911]: Why abort?

File: lib/CL/devices/cuda/pocl-ptx-gen.cc [hunk 6]
```diff
@@ -56,67 +59,173 @@
...
+                               AlignmentMapT *AlignmentMap);
+
+namespace pocl {
+extern bool isGVarUsedByFunction(llvm::GlobalVariable *GVar, llvm::Function *F);
+extern llvm::ModulePass *
+createAutomaticLocalsPass(pocl_autolocals_to_args_strategy autolocals_to_args);
+extern bool isKernelToProcess(const llvm::Function &F);
+} // namespace pocl
+
+/*
```
Comment (pjaaskel) [inline lib/CL/devices/cuda/pocl-ptx-gen.cc:95]: #if 0  commenting out is a bit nicer.

File: lib/kernel/cuda/barrier.ll [hunk 7]
```diff
@@ -10,3 +10,16 @@ entry:
   call void @llvm.nvvm.barrier0()
   ret void
 }
+
+define void @_Z22_cl_work_group_barrierj(i32) noduplicate {
+entry:
+  call void @llvm.nvvm.barrier0()
```
Comment (pjaaskel) [inline lib/kernel/cuda/barrier.ll:16]: Do we really need .ll implementations? Are there no clang builtins for these?

File: lib/kernel/cuda/atomic_add_20.ll
Comment (pjaaskel) [inline lib/kernel/cuda/atomic_add_20.ll:1]: Also here I'm wondering do we need the .ll. Cannot we really produce these from C or OpenCL C?

**Review by pjaaskel — CHANGES_REQUESTED** [review]
Great progress. Can you summarize the updates to the CUDA driver in the ReleaseNotes.rst?

File: lib/CL/devices/cuda/pocl-cuda.c (hunk 4 above)
Comment (franz) [reply lib/CL/devices/cuda/pocl-cuda.c:788]: Yes, it is after clang-format. The indentation is the same like the whole file: 2 spaces.

File: lib/CL/devices/cuda/pocl-cuda.c (hunk 5 above)
Comment (franz) [reply lib/CL/devices/cuda/pocl-cuda.c:911]: `CUDA_CHECK` macro calls abort, and always did that. I'm renaming the macro, to make it clear what it actually does.

File: lib/CL/devices/cuda/pocl-ptx-gen.cc (hunk 6 above)
Comment (franz) [reply lib/CL/devices/cuda/pocl-ptx-gen.cc:95]: fixed

**Review by pjaaskel — CHANGES_REQUESTED** [review]

**Comment by pjaaskel** [comment]
Some instability still in the LZ driver?

File: CMakeLists.txt [hunk 8]
```diff
@@ -1709,18 +1710,57 @@ if(ENABLE_CUDA)
...
+    set(CUDA_DEVICE_EXTENSIONS "${CUDA_DEVICE_EXTENSIONS} cl_khr_subgroups")
+    #set(CUDA_DEVICE_EXTENSIONS "${CUDA_DEVICE_EXTENSIONS} cl_khr_subgroup_ballot \
+    #  cl_khr_subgroup_shuffle cl_intel_subgroups cl_intel_required_subgroup_size")
+  endif()
+
+  if(CUDA_DEVICE_EXTENSIONS MATCHES "cl_khr_subgroup")
+    set(CUDA_DEVICE_FEATURES_30 "${CUDA_DEVICE_FEATURES_30} __opencl_c_subgroups")
+  endif()
+
+  option(ENABLE_CUDA_FP16 "cuda fp16" ON)
```
[the remark this answers is not in the document]
Comment (franz) [reply CMakeLists.txt:1748]: i've added documentation, FP64 i think can be enabled for all devices, FP16 is optional but defaults to ON.

File: lib/CL/devices/cuda/pocl-cuda.c (hunk 2 above)
Comment (franz) [reply lib/CL/devices/cuda/pocl-cuda.c]: fixed

File: lib/CL/devices/cuda/pocl-cuda.c (hunk 3 above)
Comment (franz) [reply lib/CL/devices/cuda/pocl-cuda.c]: fixed
